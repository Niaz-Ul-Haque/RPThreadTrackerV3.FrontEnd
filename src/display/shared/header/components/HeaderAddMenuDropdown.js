// #region imports
import React from 'react';
import PropTypes from 'prop-types';
import { NavItem, Dropdown, DropdownToggle, DropdownMenu } from 'reactstrap';
import HeaderDropdownItem from './HeaderDropdownItem';
// #endregion imports

const propTypes = {
	headerAddMenuDropdownToggle: PropTypes.func.isRequired,
	isHeaderAddMenuDropdownOpen: PropTypes.bool.isRequired,
	openUpsertCharacterModal: PropTypes.func.isRequired,
	openNewThreadModal: PropTypes.func.isRequired
};

const HeaderAddMenuDropdown = (props) => {
	const {
		headerAddMenuDropdownToggle,
		isHeaderAddMenuDropdownOpen,
		openUpsertCharacterModal,
		openNewThreadModal
	} = props;

	return (
		<NavItem>
			<Dropdown
				isOpen={isHeaderAddMenuDropdownOpen}
				data-spec="header-dropdown"
				toggle={headerAddMenuDropdownToggle}
			>
				<DropdownToggle className="nav-link dropdown-toggle">
					<i className="fas fa-plus-circle" />
				</DropdownToggle>
				<DropdownMenu data-spec="header-dropdown-menu" right className={isHeaderAddMenuDropdownOpen ? 'show' : ''}>
					<HeaderDropdownItem
						data-spec="header-dropdown-upsert-thread-link"
						onClick={() => openNewThreadModal()}
						label="Track New Thread"
					/>
					<HeaderDropdownItem
						data-spec="header-dropdown-upsert-character-link"
						onClick={() => openUpsertCharacterModal()}
						label="Add Character"
					/>
				</DropdownMenu>
			</Dropdown>
		</NavItem>
	);
};
HeaderAddMenuDropdown.propTypes = propTypes;
export default HeaderAddMenuDropdown;
