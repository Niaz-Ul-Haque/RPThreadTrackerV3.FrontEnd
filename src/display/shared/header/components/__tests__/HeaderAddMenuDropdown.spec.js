// #region imports
import React from 'react';
import { shallow } from 'enzyme';
import { getSpecWrapper } from '~/testhelpers/helpers.unit';
import HeaderAddMenuDropdown from '../HeaderAddMenuDropdown';
// #endregion imports

jest.mock('../HeaderDropdownItem', () => 'HeaderDropdownItem');

const createTestProps = propOverrides => ({
	headerAddMenuDropdownToggle: jest.fn(),
	isHeaderAddMenuDropdownOpen: true,
	logout: jest.fn(),
	openUpsertCharacterModal: jest.fn(),
	openNewThreadModal: jest.fn(),
	user: { id: '12345', userName: 'test-username' },
	...propOverrides
});

describe('rendering', () => {
	describe('snapshots', () => {
		it('should render valid snapshot', () => {
			const props = createTestProps();
			const jsx = (<HeaderAddMenuDropdown {...props} />);
			const element = shallow(jsx);
			expect(element).toMatchSnapshot();
		});
		it('should render valid snapshot when dropdown is closed', () => {
			const props = createTestProps({ isHeaderAddMenuDropdownOpen: false });
			const jsx = (<HeaderAddMenuDropdown {...props} />);
			const element = shallow(jsx);
			expect(element).toMatchSnapshot();
		});
	});
	describe('dropdown', () => {
		it('should be visible when isHeaderAddMenuDropdownOpen is true', () => {
			const props = createTestProps();
			const jsx = (<HeaderAddMenuDropdown {...props} />);
			const element = shallow(jsx);
			const menu = getSpecWrapper(element, 'header-dropdown-menu');
			expect(menu).toHaveClassName('show');
		});
		it('should be hidden when isHeaderAddMenuDropdownOpen is false', () => {
			const props = createTestProps({ isHeaderAddMenuDropdownOpen: false });
			const jsx = (<HeaderAddMenuDropdown {...props} />);
			const element = shallow(jsx);
			const menu = getSpecWrapper(element, 'header-dropdown-menu');
			expect(menu).not.toHaveClassName('show');
		});
	});
});

describe('behavior', () => {
	describe('openUpsertCharacterModal', () => {
		it('should be called when upsert character modal link is clicked', () => {
			const openUpsertCharacterModal = jest.fn();
			const props = createTestProps({ openUpsertCharacterModal });
			const jsx = (<HeaderAddMenuDropdown {...props} />);
			const element = shallow(jsx);
			const button = getSpecWrapper(element, 'header-dropdown-upsert-character-link');
			button.simulate('click');
			expect(openUpsertCharacterModal).toHaveBeenCalledTimes(1);
			expect(openUpsertCharacterModal).toHaveBeenCalledWith();
		});
	});
	describe('openNewThreadModal', () => {
		it('should be called when upsert thread modal link is clicked', () => {
			const openNewThreadModal = jest.fn();
			const props = createTestProps({ openNewThreadModal });
			const jsx = (<HeaderAddMenuDropdown {...props} />);
			const element = shallow(jsx);
			const button = getSpecWrapper(element, 'header-dropdown-upsert-thread-link');
			button.simulate('click');
			expect(openNewThreadModal).toHaveBeenCalledTimes(1);
			expect(openNewThreadModal).toHaveBeenCalledWith();
		});
	});
});
