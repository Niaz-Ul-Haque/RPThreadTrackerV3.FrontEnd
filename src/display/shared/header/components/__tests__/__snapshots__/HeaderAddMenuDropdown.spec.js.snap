// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering snapshots should render valid snapshot 1`] = `
<NavItem
  tag="li"
>
  <Dropdown
    active={false}
    addonType={false}
    data-spec="header-dropdown"
    direction="down"
    inNavbar={false}
    isOpen={true}
    nav={false}
    setActiveFromChild={false}
    toggle={[MockFunction]}
  >
    <DropdownToggle
      aria-haspopup={true}
      className="nav-link dropdown-toggle"
      color="secondary"
    >
      <i
        className="fas fa-plus-circle"
      />
    </DropdownToggle>
    <DropdownMenu
      className="show"
      data-spec="header-dropdown-menu"
      flip={true}
      right={true}
      tag="div"
    >
      <HeaderDropdownItem
        data-spec="header-dropdown-upsert-thread-link"
        label="Track New Thread"
        onClick={[Function]}
      />
      <HeaderDropdownItem
        data-spec="header-dropdown-upsert-character-link"
        label="Add Character"
        onClick={[Function]}
      />
    </DropdownMenu>
  </Dropdown>
</NavItem>
`;

exports[`rendering snapshots should render valid snapshot when dropdown is closed 1`] = `
<NavItem
  tag="li"
>
  <Dropdown
    active={false}
    addonType={false}
    data-spec="header-dropdown"
    direction="down"
    inNavbar={false}
    isOpen={false}
    nav={false}
    setActiveFromChild={false}
    toggle={[MockFunction]}
  >
    <DropdownToggle
      aria-haspopup={true}
      className="nav-link dropdown-toggle"
      color="secondary"
    >
      <i
        className="fas fa-plus-circle"
      />
    </DropdownToggle>
    <DropdownMenu
      className=""
      data-spec="header-dropdown-menu"
      flip={true}
      right={true}
      tag="div"
    >
      <HeaderDropdownItem
        data-spec="header-dropdown-upsert-thread-link"
        label="Track New Thread"
        onClick={[Function]}
      />
      <HeaderDropdownItem
        data-spec="header-dropdown-upsert-character-link"
        label="Add Character"
        onClick={[Function]}
      />
    </DropdownMenu>
  </Dropdown>
</NavItem>
`;
